Enable upstream new conformance tests

slinkydeveloper opened the pull request:
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- :gift: Add upstream new broker conformance tests to our test suite
- :bug: Fixed 405 Method not allowed conformance issue

slinkydeveloper commented:
I would love to merge https://github.com/knative-sandbox/eventing-kafka-broker/pull/911 before this one

matzew commented:
/assign

slinkydeveloper commented:
/retest

slinkydeveloper commented:
The error seems in the retries (maybe some issue with the retry asserts?):

```
    --- FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-2/Assert/[Stable/SHOULD]Conformance_When_`TriggerSpec.Delivery`_is_configured,_then_`TriggerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)
```

matzew commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/http/HttpConsumerVerticleFactory.java:
```diff
@@ -238,7 +238,7 @@ private static CircuitBreakerOptions createCircuitBreakerOptions(final DataPlane
         // TODO max failures should be configurable or, at least, set by the control plane
         .setMaxFailures(egressConfig.getRetry() * 2)
-        .setMaxRetries(egressConfig.getRetry())
+        .setMaxRetries(egressConfig.getRetry() + 1)
```
I am fine in doing the `+ 1` here. But should it work without? Perhaps reasonable to track as GH issue?

slinkydeveloper replied:
removed the +1

matzew commented on data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java:
```diff
@@ -119,6 +121,16 @@ public void handle(final HttpServerRequest request) {
+    if (request.method() != HttpMethod.POST) {
+      request.response().setStatusCode(METHOD_NOT_ALLOWED.code()).end();
+
+      logger.warn("method not allowed {}",
+        keyValue("method", request.method())
+      );
```
spec says:

> Brokers MUST reject all HTTP produce requests with a method other than POST responding with HTTP status code `405 Method Not Supported`.

Perhaps we use that as some better `WARN` text?

slinkydeveloper replied:
How about now?

matzew commented:
looking here: https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka-broker/1016/pull-knative-sandbox-eventing-kafka-broker-integration-tests/1407257040924971008

> FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-1/Assert/[Stable/SHOULD]Conformance_When_`BrokerSpec.Delivery`_is_configured,_but_not_the_specific_`TriggerSpec.Delivery`,_then_the_`BrokerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)

slinkydeveloper commented:
Interesting, now it fails on prow as well, let me try to revert the previous commit...

matzew commented on data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/RequestMapper.java:
```diff
@@ -124,7 +124,7 @@ public void handle(final HttpServerRequest request) {
       request.response().setStatusCode(METHOD_NOT_ALLOWED.code()).end();
 
-      logger.warn("method not allowed {}",
+      logger.warn("Only POST method is allowed. Method not allowed: {}",
```
:+1:

matzew reviewed (approved):
/lgtm
/approve

slinkydeveloper commented:
/hold

Let's wait until test pass consistently here

matzew commented:
> 
    --- FAIL: TestBrokerConformance/Knative_Broker_Specification_-_Control_Plane/Delivery_Spec_-_dlq-test-2/Assert/[Stable/SHOULD]Conformance_When_`TriggerSpec.Delivery`_is_configured,_then_`TriggerSpec.Delivery`_SHOULD_be_used._(Retry) (0.00s)


on the `kind` e2e, I still see the above

slinkydeveloper commented:
Yeah but it pass on prow...

pierDipi commented:
Are they failing only with eventing v0.21? Perhaps that feature wasn't available at that time? 🤷‍♂️

slinkydeveloper commented:
> Are they failing only with eventing v0.21? Perhaps that feature wasn't available at that time? man_shrugging

Uhm I don't know, retry should have been there, but maybe the overwrite trigger -> broker might be the missing bit in 0.21...

matzew commented:
should we run aganst nightly?

slinkydeveloper commented:
We run against latest as well (see 0.22). But simply, this test seems like it's failing with any < 0.22 releases

matzew commented:
I meant more if we should only run against nightly and last releaes (or last two releases)?

slinkydeveloper commented:
I think the policy of knative is to support the latest 3 major releases, that's the reason we test all these different versions.

matzew commented:
> I think the policy of knative is to support the latest 3 major releases, that's the reason we test all these different versions.

next week we cut the 0.24 :-)   so:
* 0.22
* 0.23
* 0.24
* main
done

pierDipi commented:
Let's remove eventing v0.21 since these tests are useful while we're refactoring code.

slinkydeveloper commented:
Will do in this PR now

slinkydeveloper commented:
Rebased and removed 0.21. I've also removed the new `RequestMapper` for the http method check, just because it was going to create even more conflicts with #1041 that is going to properly reorganize that code anyway.

pierDipi commented:
https://github.com/knative-sandbox/eventing-kafka-broker/pull/1016/checks?check_run_id=2917490884 `Unit Tests (1.16.x, ubuntu-latest)`
```
 git remote add upstream https://github.com/knative-sandbox/eventing-kafka-broker.git
  git fetch upstream main
  git pull --no-rebase --no-commit upstream main
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.16.5/x64
From https://github.com/knative-sandbox/eventing-kafka-broker
 * branch              main       -> FETCH_HEAD
 * [new branch]        main       -> upstream/main
From https://github.com/knative-sandbox/eventing-kafka-broker
 * branch              main       -> FETCH_HEAD
Committer identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
```

slinkydeveloper commented:
@pierDipi not this pr error at all :) I suggest we merge this one and then wait for the next actions PR fix...

pierDipi commented on .github/workflows/kind-e2e.yaml:
```diff
@@ -29,7 +29,6 @@ jobs:
           - "./third_party/eventing-latest/"
           - "./third_party/eventing-v0.23.0/"
           - "./third_party/eventing-v0.22.1/"
-          - "./third_party/eventing-v0.21.4/"
```
We need to remove this directory too

devguyio commented:
/lgtm

slinkydeveloper commented:
/unhold
